Services/communication_service.c

```diff
@@ -35,31 +53,753 @@
+        {
+            return SATR_ERROR;
+        }
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &U_is_secure, sizeof(U_is_secure));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(U_is_secure) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_FULL_STAT:
+        status =  HAL_UHF_getStatus(&U_stat.status_ctrl)
+                + HAL_UHF_getFreq(&U_stat.set.freq)
+                + HAL_UHF_getUptime(&U_stat.uptime)
+                + HAL_UHF_getPcktsOut(&U_stat.pckts_out)
+                + HAL_UHF_getPcktsIn(&U_stat.pckts_in)
+                + HAL_UHF_getPcktsInCRC16(&U_stat.pckts_in_crc16)
+                + HAL_UHF_getPIPEt(&U_stat.set.PIPE_t)
+                + HAL_UHF_getBeaconT(&U_stat.set.beacon_t)
+                + HAL_UHF_getAudioT(&U_stat.set.audio_t)
+                + HAL_UHF_getTemp(&U_stat.temperature)
+                + HAL_UHF_getLowPwr(&U_stat.low_pwr_stat)
+                + HAL_UHF_getFV(&U_stat.firmware_ver)
+                + HAL_UHF_getPayload(&U_stat.payload_size)
+                + HAL_UHF_getSecureKey(&U_stat.secure_key);
+
+        if (sizeof(U_stat) + 1 > csp_buffer_data_size())
+        {
+            return SATR_ERROR;
+        }
+        for (i = 0; i < STAT_WORD_LEN; i++)
+        {
+            U_stat.status_ctrl[i] = csp_hton32(
+                    (uint32_t) U_stat.status_ctrl[i]);
+        }
+        U_stat.set.freq = csp_hton32(U_stat.set.freq);
+        U_stat.uptime = csp_hton32(U_stat.uptime);
+        U_stat.pckts_out = csp_hton32(U_stat.pckts_out);
+        U_stat.pckts_in = csp_hton32(U_stat.pckts_in);
+        U_stat.pckts_in_crc16 = csp_hton32(U_stat.pckts_in_crc16);
+        U_stat.set.PIPE_t = csp_hton16(U_stat.set.PIPE_t);
+        U_stat.set.beacon_t = csp_hton16(U_stat.set.beacon_t);
+        U_stat.set.audio_t = csp_hton16(U_stat.set.audio_t);
+        U_stat.temperature = csp_htonflt(U_stat.temperature);
+        U_stat.low_pwr_stat = csp_hton32((uint32_t) U_stat.low_pwr_stat);
+        U_stat.firmware_ver = csp_hton32((uint32_t) U_stat.firmware_ver);
+        U_stat.payload_size = csp_hton16(U_stat.payload_size);
+        U_stat.secure_key = csp_hton32(U_stat.secure_key);
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &U_stat, sizeof(U_stat));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(U_stat) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_CALL_SIGN:
+        status =  HAL_UHF_getDestination(&U_callsign.dest)
+                + HAL_UHF_getSource(&U_callsign.src);
+
+        uhf_struct_len = 6;
+        //uhf_struct_len = U_callsign.dest.len; //if different, use this and separate loops
```

Comment:
The commented line is smarter but I got compiling errors trying to use that for defining the `dst [sizeof(unicode char) * uhf_struct_len]` since it varies and can not be used for defining the size of an array! Any suggestion?

Reply:
You're talking about dynamic memory allocation, which can be done carefully using `malloc()`, but you need to make sure to call `free()`, which I'm sure you know. That might be one solution?

Reply:
Oh, yeah, I'll give it a try. It should work.